### PR TITLE
fix: MPS/CPU compatibility for training and inference on Mac

### PR DESCRIPTION
**Pull request by jasonkneen**
## Summary

Makes nanochat fully runnable on Apple Silicon (MPS) and CPU after merging the latest upstream changes. All crashes and silent errors on non-CUDA devices are resolved.

Tested on:
- MacBook Pro M-series, torch 2.2.2 and 2.9.1
- 29/29 tests pass, 10 skipped (FA3 hardware only)

---

### `nanochat/engine.py`
- **Replace `signal.SIGALRM` with `concurrent.futures`** for calculator timeout. `SIGALRM` only works on Unix main thread — silently broken when called from FastAPI worker threads
- Guard `torch.cuda.synchronize()` behind `device_type == "cuda"`

### `nanochat/gpt.py`
- **Extract `init_rotary_embeddings()`** from `init_weights()` so checkpoint loading can restore non-persistent `cos`/`sin` buffers without re-randomising all weights
- Cast rotary buffers to `bfloat16` on CUDA only (MPS bfloat16 requires torch≥2.4; float32 used on MPS/CPU)
- **Add `F.rms_norm` fallback** for torch<2.4 (`rms_norm` was added in 2.4)
- Update `forward()` rotary dtype assertion to be device-aware

### `nanochat/optim.py`
- **`_cuda_compile()`**: skip `torch.compile(fullgraph=True)` on MPS/CPU; return function unchanged for eager execution
- **`adamw_step_fused` / `muon_step_fused`**: move 0-D CPU scalar tensors to parameter device at function start (cross-device ops crash in eager mode on MPS)
- **`muon_step_fused`**: use `bfloat16` in polar express on CUDA only; float32 on MPS/CPU
- **`_step_muon`**: replace `torch._foreach_copy_()` with `p.copy_(s)` loop on non-CUDA (`_foreach_copy_` not implemented on MPS in torch<2.4)

### `nanochat/flash_attention.py`
- **Probe SDPA for `enable_gqa` support at import time** (`enable_gqa` added in torch 2.5; `inspect.signature` raises on C builtins in older Python/torch)
- Fall back to manual KV head repetition via `repeat_interleave` when `enable_gqa` is unavailable

### `nanochat/checkpoint_manager.py`
- Call `model.init_rotary_embeddings()` instead of `model.init_weights()` after `load_state_dict()` to restore non-persistent rotary buffers without overwriting loaded weights

### `scripts/base_train.py`
- Guard `torch.compile(model)` behind `device_type == "cuda"`
- Set `mfu = None` on non-CUDA instead of computing `0/inf = 0.00%`
- Handle `mfu is None` cleanly in end-of-run report

### `tests/test_mps_compat.py` *(new)*
- 16 tests covering every fix; all pass on MPS with torch 2.2.2 and 2.9.1

### CODE REVIEW

**Comment by jwnsu** [comment]
Very nice changes for macos, thanks! Suggest we have an additional MEMORY_SIZE parameter to calculate device_batch_size and total_batch_size, eval_tokens, split_tokens.

Many macos devices have 64GB or 128GB unified memory. For example, 64GB memory can support device_batch_size=16.

**Comment by jasonkneen** [comment]
Good point! will update

**Comment by jwnsu** [comment]
Thanks, will test your revised PR and hope it will help accelerate it merging into repository -- currently running my own set of modifications, yours is more comprehensive.

**Comment by jasonkneen** [comment]
Updated! was using x86 python so updated that and currently running the training to 1000 steps

File: dev/README_MACOS.md
Comment (burtenshaw) [inline dev/README_MACOS.md:1]: I think multiple README's will be confusing to students. Why not open another PR with a separate guide on MPS use. Some guides are also shared [here](https://github.com/karpathy/nanochat/discussions) and [here](https://huggingface.co/nanochat-students).

File: dev/runcpu.sh
Comment (burtenshaw) [inline dev/runcpu.sh:1]: This is useful for dealing with all the MPS options. @jasonkneen Can you share running logs?

File: pyproject.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ dependencies = [
     "setuptools>=80.9.0",
     "tiktoken>=0.11.0",
     "tokenizers>=0.22.0",
-    "torch>=2.8.0",
+    "torch>=2.0.0,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*",
```
Comment (burtenshaw) [inline pyproject.toml]: `mps` should run on the cpu variant or `torch`. I think the use of older torch is caused by the lack of index `https://download.pytorch.org/whl/cpu`.

File: nanochat/gpt.py [hunk 2]
```diff
@@ -35,7 +35,13 @@ class GPTConfig:
 
 def norm(x):
     # Purely functional rmsnorm with no learnable params
-    return F.rms_norm(x, (x.size(-1),))
+    # Fallback for older PyTorch versions that don't have F.rms_norm
+    if hasattr(F, 'rms_norm'):
+        return F.rms_norm(x, (x.size(-1),))
+    else:
```
Comment (burtenshaw) [inline nanochat/gpt.py]: Could you try this out with the latest change son `main`? Mainly the `toml`. We should be able to use a `torch` version with `F.rms_norm`.

**Review by burtenshaw — CHANGES_REQUESTED** [review]
I took a look at this and added some inline comments. In general:
- `dev/README_MACOS.md` is a guide on nanochat for mps and I think there are better ways for the community to share guides/ tutorials than this repo.
- The mps python environment depends on older version of torch which make meaningful changes to the ml code. 
- The are a number of shell scripts that I think will confuse students. @karpathy Maybe we could deal with this in another PR by making a `recipes` directory that contains alternatives to `speedrun.sh` for alternative hardware. That will cut down on config, and maintain the 'just run it' vibe.

**Comment by jwnsu** [comment]
I've started a testing run, will share logs once it's done. It hang during Evaluating base model after lambda_openai, seems struggling with memory (on a macos 64GB).  Adding flag **--max-per-task=32** to base_eval resolved the eval issue, also  **--max-problems 100** for chat_eval.

Below is the report card for runmacos_overnight.sh:
## Summary

- Characters: 396,325
- Lines: 9,826
- Files: 49
- Tokens (approx): 99,081
- Dependencies (uv.lock lines): 2,365

| Metric          | BASE     | MID      | SFT      | RL       |
|-----------------|----------|----------|----------|----------|
| CORE            | 0.0175   | -        | -        | -        |
| ARC-Challenge   | -        | -        | 0.2700   | -        |
| ARC-Easy        | -        | -        | 0.2200   | -        |
| GSM8K           | -        | -        | 0.0000   | -        |
| HumanEval       | -        | -        | 0.0000   | -        |
| MMLU            | -        | -        | 0.3200   | -        |
| ChatCORE        | -        | -        | 0.0160   | -        |

**Comment by pavelklymenko** [comment]
👍🏻

**Comment by jasonkneen** [comment]
Will this get merged in?

File: dev/continue_training.sh [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+fi
+
+if [ "$SFT_CHECKPOINT" != "none" ]; then
+    echo "  ✓ SFT: $SFT_CHECKPOINT"
+else
+    echo "  ✗ SFT: Not found"
+fi
+echo ""
+
+# Determine what to do
```
Comment (karpathy) [inline dev/continue_training.sh:120]: This script is being very clever and it's bloating the file a lot. It seems like it's designed to be run over and over and reliably push the whole pipeline through. I was thinking of these scripts as more partly documentation, and partly an example of a thing you can run and wait a long time to get result. All of this logic is complexifying it quite a bit and makes it harder to read and understand what the sequence of steps is. I'm ok with some of the "autodetect hyperparams" up top. But I don't think all of the other logic here is a good idea unless you can justify.

File: nanochat/gpt.py [hunk 4]
```diff
@@ -35,7 +35,13 @@ class GPTConfig:
 
 def norm(x):
     # Purely functional rmsnorm with no learnable params
-    return F.rms_norm(x, (x.size(-1),))
+    # Fallback for older PyTorch versions that don't have F.rms_norm
+    if hasattr(F, 'rms_norm'):
```
Comment (karpathy) [inline nanochat/gpt.py:45]: Surprised this wouldn't exist. Are we trying to support way too old version of pytorch or so

File: nanochat/muon.py [hunk 5]
```diff
@@ -19,7 +18,11 @@ def zeropower_via_newtonschulz5(G: Tensor, steps: int) -> Tensor:
     """
     assert G.ndim >= 2 # batched Muon implementation by @scottjmaddox, and put into practice in the record by @YouJiacheng
     a, b, c = (3.4445, -4.7750,  2.0315)
-    X = G.bfloat16()
+    # Use bfloat16 on CUDA for speed, float32 on MPS/CPU (MPS doesn't support bfloat16)
+    if G.device.type == "cuda":
+        X = G.bfloat16()
+    else:
```
Comment (karpathy) [inline nanochat/muon.py]: possibly able to skip the else block here.

File: pyproject.toml [hunk 6]
```diff
@@ -14,7 +14,7 @@ dependencies = [
     "setuptools>=80.9.0",
     "tiktoken>=0.11.0",
     "tokenizers>=0.22.0",
-    "torch>=2.8.0",
+    "torch>=2.0.0,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*",  # Excludes versions without macOS arm64 wheels
```
Comment (karpathy) [inline pyproject.toml]: Why not leave as before?

File: pyproject.toml [hunk 7]
```diff
@@ -44,19 +44,6 @@ python_files = ["test_*.py"]
 python_classes = ["Test*"]
 python_functions = ["test_*"]
 
-# target torch to cuda 12.8
```
Comment (karpathy) [inline pyproject.toml]: Wait isn't all of this needed to correctly route the indices

File: pyproject.toml (hunk 6 above)
Comment (jubruckne) [inline pyproject.toml]: I'm running torch 2.9 so wheels seem to have been added.

File: README.md [hunk 8]
```diff
@@ -95,7 +95,38 @@ And a bit more about computing environments that will run nanochat:
...
+- 128GB: batch_size=16, total_batch=16384 (16× faster)
+- 64GB: batch_size=8, total_batch=8192 (8× faster)
+- 32GB: batch_size=4, total_batch=4096 (4× faster)
+- <32GB: batch_size=1, total_batch=1024 (baseline)
+
+**Quick start on Mac:**
+```bash
+bash dev/runcpu.sh              # Quick test (~15-30 min on 128GB Mac)
+bash dev/runmac_overnight.sh    # Better quality (~2-3 hours on 128GB Mac)
+bash dev/continue_training.sh   # Resume interrupted training
```
Comment (karpathy) [inline README.md]: I don't love the continue training script, per my comment also above.

**Comment by karpathy** [comment]
@jasonkneen  Thank you I want to merge most of this. I am most suspicious of all the uv / dependency changes. I'd prefer we not touch the dependencies unless we have to. And I think I don't wish to support continue training, it feels too bloating.

**Comment by karpathy** [comment]
One more question on my mind - how come all of this worked on my macbook fine without all of the changes?

**Comment by jasonkneen** [comment]
It did? That's awkward 🧐 I can check on the uv changes and see if it's necessary

**Comment by svlandeg** [comment]
Hmmm wait. So this PR got closed when https://github.com/karpathy/nanochat/pull/88 was merged (Oct 21) and I subsequently removed the feature branch `cpu-mps-dev` I guess to clean up. That automatically closed this PR on Oct 30 because it targets that (merged) feature branch. But there's plenty of review discussion still going on here from within that time period Oct 21-30 so maybe this was closed prematurely by accident, should target `master` instead and should get another look at. 🤔

**Comment by svlandeg** [comment]
@jasonkneen : are you still around to have a look at the reviewer comments and to update this PR for the latest on `master`?

**Comment by jasonkneen** [comment]
sure I can take a look

File: pyproject.toml (hunk 6 above)
Comment (icetraxx) [reply pyproject.toml]: All of those versions do contain arm64 wheels:

https://pypi.org/project/torch/2.3.0/#files
https://pypi.org/project/torch/2.4.0/#files
https://pypi.org/project/torch/2.5.0/#files
https://pypi.org/project/torch/2.6.0/#files
https://pypi.org/project/torch/2.7.0/#files
https://pypi.org/project/torch/2.8.0/#files
https://pypi.org/project/torch/2.9.1/#files

**Comment by jasonkneen** [comment]
Apologies - I am looking at this now

**Comment by svlandeg** [comment]
@jasonkneen: there are plenty of open reviewer's comments, could you look into those as well?

**Comment by jasonkneen** [comment]
will do
